mabwiser/treebandit.py

```diff
@@ -0,0 +1,173 @@
+            except AttributeError:
+                # If arm's tree has not been fitted
+                self.arm_to_tree[arm].fit(contexts[decisions == arm], rewards[decisions == arm])
+                unfitted_arms.append(arm)
+
+    def predict(self, contexts: np.ndarray = None) -> Arm:
+
+        return self._parallel_predict(contexts, is_predict=True)
+
+    def predict_expectations(self, contexts: np.ndarray = None) -> Dict[Arm, Num]:
+
+        return self._parallel_predict(contexts, is_predict=False)
+
+    def _fit_arm(self, arm: Arm, decisions: np.ndarray, rewards: np.ndarray, contexts: Optional[np.ndarray] = None):
+
+        # Create dataset for the given arm
+        arm_contexts = contexts[decisions == arm]
+        arm_rewards = rewards[decisions == arm]
+
+        # Check that the dataset for the given arm is not empty
+        if arm_contexts.size != 0:
+
+            # Fit decision tree of the arm with this dataset
+            self.arm_to_tree[arm].fit(arm_contexts, arm_rewards)
+
+            # For each leaf, keep a list of rewards
+            # DecisionTreeClassifier's apply() method returns the indices of the nodes in the tree
+            # that the specified contexts lead to. Therefore, the indices returned are not necessarily
+            # consecutive/follow numerical order, but are always representative of leaf nodes.
+            leaf_indices = self.arm_to_tree[arm].apply(arm_contexts)
+
+            # Use set() to create a list of unique indices
+            # These indices represent the leaves reached via the given contexts
+            unique_leaf_indices = set(leaf_indices)
+
+            for index in unique_leaf_indices:
+                # Get rewards list for each leaf
+                self.arm_to_rewards[arm][index] = arm_rewards[leaf_indices == index]
+
+    def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
+                          seeds: Optional[np.ndarray] = None, start_index: Optional[int] = None) -> List:
+
+        # Get local copy of arm_to_tree, arm_to_expectation, arm_to_rewards, and arms
+        # to minimize communication overhead between arms (processes) using shared objects
+        arm_to_tree = deepcopy(self.arm_to_tree)
+        arm_to_rewards = deepcopy(self.arm_to_rewards)
+        arm_to_expectation = deepcopy(self.arm_to_expectation)
+        arms = deepcopy(self.arms)
+
+        # Create an empty list of predictions
+        predictions = [None] * len(contexts)
+        for index, row in enumerate(contexts):
+            # Each row needs a separately seeded rng for reproducibility in parallel
+            rng = np.random.RandomState(seed=seeds[index])
+
+            for arm in arms:
+
+                # Copy the row rng to the deep copied model in arm_to_tree
+                arm_to_tree[arm].rng = rng
+
+                # If there was prior data for this arm, get expectation for arm
+                if arm_to_rewards[arm]:
```

Comment:
The dictionary is created from self.arms, not from the set of arms that have training data. What is the behavior for an arm that doesn't have a fitted model?

Reply:
TreeBandit recognizes the arm and keeps an unfitted decision tree and a dictionary for leaf rewards for it. However, upon receiving a context to predict for, it likely wouldn't choose this arm since it's expected reward is 0.

Reply:
In the fit function you call self.arm_to_rewards = {arm: dict() for arm in self.arms}. So 'arm' will always be available as a key in self.arm_to_rewards regardless of whether or not the arm has observed rewards and this line will always return true

Reply:
I just checked again, I believe when a numpy array is empty 'if nparray' evaluates to False.

Reply:
I see